### PR TITLE
fix(codegen): point maskingConstId_ at samv-maskvalue, not constant 0

### LINKED ISSUE #4390: native `mean` reduction ignores `coordinateMasking_`
### What happened?

For a stick-dim reduction with non-stick-aligned size (i.e. padding is
present), the SuperDSC frontend (torch-spyre) emits **identical**
`coordinateMasking_` / `maskingConstId_` / `samv-maskvalue` directives for
`sum` and `mean`. The `sum` backend op honors this directive and produces
correct results. The `mean` backend op does not: uninitialized/garbage
values in the padding lanes leak into the reduction, producing wrong
answers whenever the reduced size isn't a multiple of the 64-element
fp16 stick.

This was investigated from the torch-spyre side under
[torch-spyre#1706](https://github.com/torch-spyre/torch-spyre/issues/1706)
and [torch-spyre#2719](https://github.com/torch-spyre/torch-spyre/issues/2719).
The frontend-side scaling (`1/N`) and masking metadata were both verified
correct (see "What was ruled out" below), narrowing the defect to the
native `mean` op inside `dxp_standalone`.

### What did you expect to happen?

The correct result is computed.

## Reproducers

- `repro_sum_correct.py` — `torch.sum(x, dim=-1, keepdim=True)`, shape
  `(3, 7, 9)`, fp16. **Passes.**
- `repro_mean_wrong.py` — identical shape/dim/dtype, `torch.mean` instead
  of `torch.sum`. **Fails.**

Both scripts:

1. Run on an all-zeros input first. This is the clearest control: the
   correct answer is *exactly* 0.0 regardless of what garbage sits in
   the padding lanes, so any nonzero result proves padding leakage.
2. Run on a random input and compare against the CPU (eager) reference
   with `torch.testing.assert_close`.

Run with:

```bash
TORCH_COMPILE_DEBUG=1 TORCH_LOGS="+inductor" SENCORES=1 python3 repro_sum_correct.py
TORCH_COMPILE_DEBUG=1 TORCH_LOGS="+inductor" SENCORES=1 python3 repro_mean_wrong.py
```

### Observed output

`repro_sum_correct.py`:

```
=== all-zeros input ===
spyre sum: [0.0, 0.0, 0.0, ...]   (21 zeros)
PASS: sum of all-zeros input is exactly 0
...
PASS: sum matches CPU reference
```

`repro_mean_wrong.py`:

```
=== all-zeros input ===
spyre mean: [0.6796875, 0.6796875, 0.6796875, ...]   (21 copies of the same nonzero garbage value)
expected  : [0.0, 0.0, 0.0, ...]
FAIL: mean of all-zeros input is NOT 0 -- padding-lane garbage leaked into the reduction
...
FAIL: mean does not match CPU reference
Mismatched elements: 21 / 21 (100.0%)
Greatest absolute difference: 0.68017578125 at index (0, 5, 0) (up to 0.01 allowed)
```

Note the mean error (`0.6797`) is consistent across every output element —
it is a fixed per-kernel-invocation garbage sum divided by N=9, not
data-dependent noise. This is exactly what you'd expect from a fixed
amount of uninitialized/garbage memory in the unmasked padding lanes
being included in the accumulation.

## Artifacts

For each of the two runs, the full compiler bundle is captured under
`artifacts/<name>/`:

- `sdsc_0.json` / `sdsc_0.pretty.json` — the SuperDSC JSON descriptor
  handed to `dxp_standalone` (this is what `bundle.mlir` references via
  `sdsc_filename="sdsc_0.json"`)
- `bundle.mlir` — the MLIR wrapper (identical between sum/mean; it's a
  thin `sdsc_execute` dispatcher, all real content is in `sdsc_0.json`)
- `spyreCodeDir/spyrecode.json`, `spyrecode_pretty.json` — post-compile
  device code descriptor
- `spyreCodeDir/init_binary.bin` — compiled binary blob
- `inductor_debug/` — Inductor-side debug artifacts (FX graph,
  `ir_post_fusion.txt`, generated `output_code.py`) for reference; not
  part of the backend compiler's input, but useful for tracing how the
  SDSC JSON was derived

Directories:

```
artifacts/sum_correct/   <- torch.sum,  correct
artifacts/mean_wrong/    <- torch.mean, wrong
```

## Key diff: `sdsc_0.json`

The only substantive differences between the two SDSC descriptors
(`diff artifacts/sum_correct/sdsc_0.pretty.json artifacts/mean_wrong/sdsc_0.pretty.json`)
are:

1. `opFuncName`: `"sum"` vs `"mean"`
2. `mean` has one extra `constantInfo_` entry, `scaling_factor` (the
   `1/N` multiplier applied after reduction — verified correct by
   decoding its `SEN169_FP16` bit pattern)

Everything relevant to padding/masking is **byte-identical**:

| Field | sum | mean |
|---|---|---|
| `coordinateMasking_` | `{"out": [[9, 55]]}` | `{"out": [[9, 55]]}` |
| `maskingConstId_` | `0` | `0` |
| `samv-maskvalue` constant | `0` | `0` |

i.e. both ops are told: "coordinates 9..63 of the `out` dim are padding;
mask them to value 0 (constant id 0) before reducing." `sum` does this.
`mean` does not.

## What was ruled out (frontend-side)

- **Scaling factor / `1/N` math**: `scaling_factor = 1/9 = 0.1111...`,
  encoded via `encode_constant` into the on-device `SEN169_FP16` format
  (a non-IEEE754 1-sign/6-exponent/9-mantissa layout, bias 31). Decoding
  the emitted bit pattern (`14222`) confirms it round-trips to
  `1/9` correctly.
- **Coordinate mask range math**: `_get_coordinate_mask` in
  `torch_spyre/_inductor/codegen/superdsc.py` computes the same
  `[padded_size - padding, padding]` range for both ops — confirmed
  identical in the JSON above.
- **Which reduction size triggers it**: reproduced across a sweep of
  reduction sizes 2-63 (padding 1-62); the all-zeros garbage leak is
  present at *every* non-stick-aligned size, shrinking smoothly as
  padding shrinks (consistent with `garbage_sum / N`), and disappears
  entirely when the reduced size is a multiple of 64 (no padding) or
  when the reduced dim is not the stick dim at all.

All of this points at the native `mean` kernel in `dxp_standalone` not
applying the coordinate mask before accumulating, while `sum`'s kernel
does.

bundle.mlirs and sdsc.jsons in attached tar file.

[artifacts.tgz](https://github.com/user-attachments/files/31978673/artifacts.tgz)

### Any environmental details we need to know?

No.

### Anything else we need to know?

_No response_

### Relevant log output

```Shell

```

### PR DESCRIPTION
**Pull request by lupalby**
Fixes #4390.

The defect is in the SuperDSC frontend, not the native `mean` kernel — the backend honors `coordinateMasking_` for both `sum` and `mean`, but is told to mask with the wrong value.

`compute_ops.py` hardcoded `"maskingConstId_": 0`. Constant ids are positions in the `constants` dict, assigned by insertion order in `generate_constant_info`, and `superdsc.py` appends `samv-maskvalue` *after* any op-supplied constants. `sum` carries no others, so index 0 was correct by accident. `mean` inserts `scaling_factor` first, which shifts `samv-maskvalue` to index 1 while `maskingConstId_` stayed pinned at 0 — so the padding lanes were seeded with 1/N instead of the mask value, and that leaked into the reduction.

This also explains why the reported error was a fixed value rather than data-dependent noise: it is a deterministic constant, not uninitialized garbage. For the reported case (reduce 9 → 55 padding lanes holding 1/9, then scaled by 1/9) the predicted result is 0.678681 against 0.679688 observed, the difference being fp16 accumulation rounding.

## Fix

Record the index where `samv-maskvalue` is actually inserted and carry it on `SDSCSpec.masking_const_id`. `parse_op_spec` is the only place that inserts the constant and the only production `SDSCSpec` construction site.

## Verification

Reproducers written from scratch, run on hardware, `(3, 7, 9)` fp16, `dim=-1`:

| | before | after |
|---|---|---|
| `mean`, all-zeros control | 0.679688 ✗ | 0.0 ✓ |
| `mean` vs CPU eager (random) | 21/21 mismatched ✗ | pass ✓ |
| `sum`, both checks | pass | pass (unchanged) |

The all-zeros case is the decisive control: the correct answer is exactly 0 regardless of what sits in the padding lanes.

Reduction-size sweep (8, 9, 17, 31, 63, 64, 65, 127, 128) is clean; previously every non-multiple-of-64 leaked. `tests/inductor/` `test_codegen.py`, `test_coarse_tiling.py`, `test_provenance.py`: 440 passed.

`TestMaskingConstId` in `test_codegen.py` pins the id against the real `generate_constant_info`, and pins the assumption the scheme rests on — that `constants` iterates in insertion order, not sorted. Note `"samv-maskvalue"` sorts *before* `"scaling_factor"`, so a switch to sorted ordering would silently reintroduce this bug.

### CODE REVIEW

**Review by avery-blanchard — APPROVED** [review]
looks good to me
